client/consensus/manual-seal/src/rpc.rs

```diff
@@ -0,0 +1,191 @@
+
+//! RPC interface for the ManualSeal Engine.
+use consensus_common::ImportedAux;
+use jsonrpc_core::{Error, ErrorCode};
+use jsonrpc_derive::rpc;
+use futures::{
+	channel::{
+		mpsc::{self, SendError},
+		oneshot,
+	},
+	TryFutureExt,
+	SinkExt
+};
+use serde::{Deserialize, Serialize};
+
+type FutureResult<T> = Box<dyn jsonrpc_core::futures::Future<Item = T, Error = Error> + Send>;
+type Sender<T> = Option<oneshot::Sender<std::result::Result<T, crate::Error>>>;
+
+/// message sent by rpc to the background authorship task
+pub enum EngineCommand<Hash> {
+	/// Tells the engine to propose a new block
+	///
+	/// if create_empty == true, it will create empty blocks if there are no transactions
+	/// in the transaction pool
+	SealNewBlock {
+		create_empty: bool,
+		parent_hash: Option<Hash>,
+		sender: Sender<CreatedBlock<Hash>>,
+	},
+	/// Tells the engine to finalize the block with the supplied hash
+	FinalizeBlock {
+		hash: Hash,
+		sender: Sender<()>,
+	}
+}
+
+#[rpc]
+pub trait ManualSealApi<Hash> {
+	/// Instructs the manual-seal background task to create a new block
+	#[rpc(name = "engine_createBlock")]
+	fn create_block(
+		&self,
+		create_empty: bool,
+		parent_hash: Option<Hash>
+	) -> FutureResult<CreatedBlock<Hash>>;
+
+	/// Instructs the manual-seal background task to finalize a block
+	#[rpc(name = "engine_finalizeBlock")]
+	fn finalize_block(
+		&self,
+		hash: Hash,
+	) -> FutureResult<()>;
+}
+
+/// A struct that implements the [`ManualSealApi`].
+pub struct ManualSeal<Hash> {
+	import_block_channel: mpsc::Sender<EngineCommand<Hash>>,
+}
+
+/// return type of `engine_createBlock`
+#[derive(Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub struct CreatedBlock<Hash> {
+	pub hash: Hash,
+	pub aux: ImportedAux
+}
+
+impl<Hash> ManualSeal<Hash> {
+	/// Create new `ManualSeal` with the given reference to the client.
+	pub fn new(import_block_channel: mpsc::Sender<EngineCommand<Hash>>) -> Self {
+		Self { import_block_channel }
+	}
+}
+
+impl<Hash: Send + 'static> ManualSealApi<Hash> for ManualSeal<Hash> {
+	fn create_block(
+		&self,
+		create_empty: bool,
+		parent_hash: Option<Hash>
+	) -> FutureResult<CreatedBlock<Hash>> {
+		let mut sink = self.import_block_channel.clone();
+		let future = async move {
+			let (sender, receiver) = oneshot::channel();
+			let result = sink.send(
+				EngineCommand::SealNewBlock {
+					create_empty,
+					parent_hash,
+					sender: Some(sender),
+				}
+			).await;
+			if let Err(e) = result {
+				return Err(map_error(e))
+			};
+
+			match receiver.await {
+				// all good
+				Ok(Ok(block)) => Ok(block),
+				// error from the authorship task
+				Ok(Err(e)) => {
+					Err(Error {
+						code: ErrorCode::ServerError(500),
+						message: format!("{}", e),
+						data: None
+					})
+				}
+				// channel has been dropped
+				Err(_) => {
+					Err(Error {
+						code: ErrorCode::ServerError(500),
+						message: "Server is shutting down".into(),
```

Comment:
Where did you draw this error message and code from?

Maybe we can rephrase to "Consensus process terminating", if the only reason that this can happen is that the `manual_seal` future is terminating.